simperby/src/node/storage.rs

```diff
@@ -0,0 +1,176 @@
+use crate::*;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum HistoryStorageError {
+    #[error("invalid operation: {0}")]
+    InvalidOperation(String),
+    #[error("storage error: {0}")]
+    StorageError(simperby_kv_storage::Error),
+    /// When an essential item is missing or corrupted.
+    #[error("malformed storage: {0}")]
+    StorageIntegrityError(String),
+}
+
+type Error = HistoryStorageError;
+
+impl From<simperby_kv_storage::Error> for Error {
+    fn from(e: simperby_kv_storage::Error) -> Self {
+        Error::StorageError(e)
+    }
+}
+
+impl From<Error> for SimperbyError {
+    fn from(e: Error) -> Self {
+        match e {
+            HistoryStorageError::InvalidOperation(e) => SimperbyError::InvalidOperation(e),
+            HistoryStorageError::StorageError(e) => SimperbyError::StorageError(e),
+            HistoryStorageError::StorageIntegrityError(e) => {
+                SimperbyError::StorageIntegrityError(e)
+            }
+        }
+    }
+}
+
+fn create_block_key(height: BlockHeight) -> Hash256 {
+    Hash256::hash(format!("block-{}", height).as_bytes())
+}
+
+fn create_last_finalized_height_key() -> Hash256 {
+    Hash256::hash("last-finalized-height".as_bytes())
+}
+
+fn create_last_finalized_block_proof_key() -> Hash256 {
+    Hash256::hash("last-finalized-block-finalization-proof".as_bytes())
+}
+
+fn initialization_flag_key() -> Hash256 {
+    Hash256::hash("initialization-flag".as_bytes())
+}
+
+async fn get_json_from_storage<T>(storage: &dyn KVStorage, key: Hash256) -> Result<T, Error>
+where
+    T: serde::de::DeserializeOwned,
+{
+    let value = storage.get(key).await?;
+    serde_json::from_slice(&value).map_err(|e| Error::StorageIntegrityError(e.to_string()))
+}
+
+async fn set_json_to_storage<T>(
```

Comment:
How about using "store" instead of "set"?

Reply:
fixed to `insert_or_update_json_to_storage`

Reply:
cool